MaxText/layers/attentions.py

```diff
@@ -277,9 +301,13 @@ def apply_attention_dot(
     # QK Product, a.k.a `attn_weights`: [batch, num_kv_heads, num_query_heads_per_kv_head, q_length, kv_length]
     attn_weights = self.qk_product(query, key)
 
+    # convert attn_weights to jnp.float32 for 2 reasons
+    #   1) stability for the softmax function
```

Comment:
Was this important for convergence? Very interesting.

Reply:
Will have some datapoints by tomorrow.

Reply:
Found very tiny improvement:
  * ~1e-4 better in eval_loss i.e equivalent of 1 train_step in average.
Will disable it by default.

* bf16 with flash attention: 
  * enable fp32[logs](https://pantheon.corp.google.com/logs/query;query=resource.type%3D%22k8s_container%22%0Aresource.labels.location%3D%22us-east5%22%0Aresource.labels.namespace_name%3D%22default%22%0Aresource.labels.project_id%3D%22tpu-prod-env-multipod%22%0Aresource.labels.container_name%3D%22jax-tpu%22%0Aresource.labels.cluster_name%3D%22v5p-1024-mlperf%22%0Aresource.labels.pod_name%3D%22lizhiyu-240121-151924-slice-job-0-121-mqgtn%22;pinnedLogId=2024-01-21T18:35:55.292791107Z%2Fgu036fyr8hufrak1;cursorTimestamp=2024-01-21T15:20:57.265513534Z;startTime=2024-01-21T12:39:46.637Z;endTime=2024-01-21T18:39:46.637Z?e=13802955&mods=allow_workbench_image_override&project=tpu-prod-env-multipod&pli=1): step 276: eval_loss: 2.69041; step 288: eval_loss: 2.68903
  * disable fp32[logs](https://pantheon.corp.google.com/logs/query;query=resource.type%3D%22k8s_container%22%0Aresource.labels.project_id%3D%22tpu-prod-env-multipod%22%0Aresource.labels.location%3D%22us-east5%22%0Aresource.labels.container_name%3D%22jax-tpu%22%0Aresource.labels.pod_name%3D%22lizhiyu-240124-000827-bf16-slice-job-0-12-bxjw6%22%0Aresource.labels.namespace_name%3D%22default%22%0Aresource.labels.cluster_name%3D%22v5p-1024-mlperf%22;pinnedLogId=2024-01-24T03:22:53.862893827Z%2Fa739ijfzsvl0fmz3;cursorTimestamp=2024-01-24T03:23:12.750211414Z;startTime=2024-01-23T21:21:14.159Z;endTime=2024-01-24T06:21:14.159Z?e=13802955&mods=allow_workbench_image_override&project=tpu-prod-env-multipod&pli=1) step 276: eval_loss: 2.69051; step 288: eval_loss: 2.68915
 * int8 with mha
   * enable fp32 [logs](https://pantheon.corp.google.com/logs/query;query=resource.type%3D%22k8s_container%22%0Aresource.labels.container_name%3D%22jax-tpu%22%0Aresource.labels.cluster_name%3D%22v5p-1024-mlperf%22%0Aresource.labels.pod_name%3D%22lizhiyu-240124-033939-mha-int8-slice-job-0-23-55z2d%22%0Aresource.labels.location%3D%22us-east5%22%0Aresource.labels.project_id%3D%22tpu-prod-env-multipod%22%0Aresource.labels.namespace_name%3D%22default%22;pinnedLogId=2024-01-24T06:11:06.714867069Z%2Fi3uh8chshd3d5h3d;cursorTimestamp=2024-01-24T06:11:06.715429908Z;duration=PT1H?e=13802955&mods=allow_workbench_image_override&project=tpu-prod-env-multipod&pli=1): step 276: eval_loss 2.69114; step 288: eval_loss: 2.68990
   * disable fp32 [logs](https://pantheon.corp.google.com/logs/query;query=resource.type%3D%22k8s_container%22%0Aresource.labels.namespace_name%3D%22default%22%0Aresource.labels.cluster_name%3D%22v5p-1024-mlperf%22%0Aresource.labels.pod_name%3D%22lizhiyu-240123-234522-mha-int8-slice-job-0-40-62kds%22%0Aresource.labels.project_id%3D%22tpu-prod-env-multipod%22%0Aresource.labels.location%3D%22us-east5%22%0Aresource.labels.container_name%3D%22jax-tpu%22;pinnedLogId=2024-01-24T02:16:37.471613102Z%2Fu7i1phn35jef81u8;cursorTimestamp=2024-01-24T02:13:57.281457907Z;startTime=2024-01-24T00:18:30.928Z;endTime=2024-01-24T06:18:30.928Z?e=13802955&mods=allow_workbench_image_override&project=tpu-prod-env-multipod&pli=1): step 276: eval_loss 2.69122; step 288: eval_loss: 2.68997

Reply:
Done.